Bottom: 321fc85e019590580a44e636f0ce8f6b042578d3
Top:    8f2e2e7be2a40612a8094174f3600312aa39f98b
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-04-16 21:04:15 +0800

Refresh of tuntap-split-out-xdp

---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index cd01260..b111643 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1642,8 +1642,10 @@ static struct sk_buff *tun_do_xdp(struct tun_struct *tun,
 	}
 
 	skb = build_skb(xdp->data_hard_start, buflen);
-	if (!skb)
-		return ERR_PTR(-ENOMEM);
+	if (!skb) {
+		skb = ERR_PTR(-ENOMEM);
+		goto err_xdp;
+	}
 
 	skb_reserve(skb, pad - delta);
 	skb_put(skb, len + delta);
@@ -1651,6 +1653,7 @@ static struct sk_buff *tun_do_xdp(struct tun_struct *tun,
 out:
 	return skb;
 err_xdp:
+	put_page(virt_to_head_page(xdp->data_hard_start));
 	return ERR_PTR(-err);
 }
 
@@ -1718,6 +1721,7 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 
 	skb = build_skb(buf, buflen);
 	if (!skb) {
+		put_page(alloc_frag->page);
 		skb = ERR_PTR(-ENOMEM);
 		goto out;
 	}
@@ -1731,8 +1735,6 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 	return skb;
 
 err_xdp:
-	alloc_frag->offset -= buflen;
-	put_page(alloc_frag->page);
 	this_cpu_inc(tun->pcpu_stats->rx_dropped);
 out:
 	rcu_read_unlock();
